### DISCUSSION TITLE
Cannot use map function inside evaluate

### DISCUSSION THREAD

**Discussion by WilcovS** [josdejong/mathjs#3238] [Q&A] [answered]
I'm currently evaluating various functions using `evaluate`, but I can't figure out how to use functions that accept functions as their arguments.

Example:
`math.evaluate('map([1, 2, 3], x => x+x)') // expect [2,4,6]`
But instead, I get: 
> Value expected (char 19)

Any mathjs function that accepts a callback or function as an argument seems to not work, but all the functions that don't, do.
this works for example:
`math.evaluate('add(1,2,3)') // returns 6`

Am I missing something, or is this a limitation of the library?

**Comment by dvd101x** [comment, marked as answer]
Hi,

The syntax for arrow functions in javascript is not the syntax for functions in the parser.

This is not a function `x => x+x`, instead try `f(x) = x+x` personally if I wont be using `f` later I tend to use  `_(x) = x+x` that will make `_` into a function but that reminds me I wont be using that later.

You could try the following and get the expected result.
```js
math.evaluate('map([1, 2, 3], _(x) = x+x)')
```

Here is a reference for defining functions.
https://mathjs.org/docs/expressions/syntax.html#differences-from-javascript

**Reply by WilcovS** [reply]
Thank you! I had missed that page. Perhaps it's a good idea to link to it from the evaluate doc?
https://mathjs.org/docs/reference/functions/evaluate.html

**Reply by josdejong** [reply]
Good idea. Addressed via 61c5d07ae16a4a1da57df3e7c1261e5252ade689

**Reply by dvd101x** [reply]
Somehow this also works.

```js
math.evaluate('map([1, 2, 3], x+x)')
```

Jos, I recall at some point you mentioned this was undocumented. If it's still undocumented, what is a good way to document it?

In some curious cases this notation causes errors.
```js
math.evaluate('map([1, 2, 3], val + idx)') // error
math.evaluate('map([1, 2, 3], x)')         // error
```

**Reply by josdejong** [reply]
mathjs tries to detect a single undefined variable and assumes that will be the variable to be used to iterate over the matrix values. In your example `math.evaluate('map([1, 2, 3], val + idx)')` there are two undefined variables. The first can be used to iterate over, but the second remains undefined (when evaluating this, you'll get an error that gives a clue: "Undefined symbol idx"). Does that make sense?

This is handled by the `compileInlineExpression` function here: https://github.com/josdejong/mathjs/blob/develop/src/expression/transform/map.transform.js#L29-L30

I did look around in the documentation but indeed I think the documentation is missing. It would be neat to add a section about map and forEach on the [Syntax page](https://mathjs.org/docs/expressions/syntax.html). Is that something you would like to pick up @dvd101x ? (if not I'll do it).

**Reply by dvd101x** [reply]
Hi @josdejong, I will do it.
